### ISSUE TITLE
Incorrect 'Implicit return in function which does not return' when using 'contextlib.contextmanager'

### ISSUE THREAD

**Issue by tomwatson1024** [python/mypy#8401]
Consider the following snippet.

```python
import contextlib
import sys
from typing import Iterator, NoReturn

def foo() -> NoReturn:
    with bar():
        pass
    sys.exit(0)


@contextlib.contextmanager
def bar() -> Iterator[None]:
    yield None
```

Running mypy produces

```python
error: Implicit return in function which does not return  [misc]
    def foo() -> NoReturn:
    ^
```

Hand-writing the context manager (as follows) does not produce the same error.

```python
class Bar:
    def __enter__(self) -> None:
        return None

    def __exit__(self, *args: Any) -> None:
        pass
```

I think the error in the contextlib case is a false positive.

mypy version `0.761` and `0.770+dev.a07dbd00638c35729e30dfc4ae441a8985183563` (current master)
Python version `3.6.9`
flags `--ignore-missing-imports --check-untyped-defs --disallow-untyped-defs --no-implicit-optional --warn-unused-ignores`

May be related to https://github.com/python/mypy/issues/8129.

**Comment by ilevkivskyi** [comment]
Yes, moving the definition of `bar()` _before_ definition of `foo()` fixes the error, which means this is a duplicate of https://github.com/python/mypy/issues/8129
